packages/ckeditor5-source-editing/src/sourceediting.js

```diff
@@ -0,0 +1,337 @@
+		this.set( 'isSourceEditingMode', false );
+
+		/**
+		 * The element replacer instance used to replace the editing roots with the wrapper elements containing the document source.
+		 *
+		 * @private
+		 * @member {module:utils/elementreplacer~ElementReplacer}
+		 */
+		this._elementReplacer = new ElementReplacer();
+
+		/**
+		 * Maps all root names to wrapper elements containing the document source.
+		 *
+		 * @private
+		 * @member {Map.<String,HTMLElement>}
+		 */
+		this._replacedRoots = new Map();
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const t = editor.t;
+
+		editor.ui.componentFactory.add( 'sourceEditing', locale => {
+			const buttonView = new ButtonView( locale );
+
+			buttonView.set( {
+				label: t( 'Source' ),
+				icon: sourceEditingIcon,
+				tooltip: true,
+				withText: true,
+				class: 'ck-source-editing-button'
+			} );
+
+			buttonView.bind( 'isOn' ).to( this, 'isSourceEditingMode' );
+
+			// The button should be disabled if one of the following conditions is met:
+			buttonView.bind( 'isEnabled' ).to(
+				this, 'isEnabled',
+				editor, 'isReadOnly',
+				editor.plugins.get( PendingActions ), 'hasAny',
+				( isEnabled, isEditorReadOnly, hasAnyPendingActions ) => {
+					// (1) The plugin itself is disabled.
+					if ( !isEnabled ) {
+						return false;
+					}
+
+					// (2) The editor is in read-only mode.
+					if ( isEditorReadOnly ) {
+						return false;
+					}
+
+					// (3) Any pending action is scheduled. It may change the model, so modifying the document source should be prevented
+					// until the model is finally set.
+					if ( hasAnyPendingActions ) {
+						return false;
+					}
+
+					return true;
+				}
+			);
+
+			this.listenTo( buttonView, 'execute', () => {
+				this.isSourceEditingMode = !this.isSourceEditingMode;
+			} );
+
+			return buttonView;
+		} );
+
+		// Currently, the plugin handles the source editing mode by itself only for the classic editor. To use this plugin with other
+		// integrations, listen to the `change:isSourceEditingMode` event and act accordingly.
+		if ( this._isAllowedToHandleSourceEditingMode() ) {
+			this.on( 'change:isSourceEditingMode', ( evt, name, isSourceEditingMode ) => {
+				if ( isSourceEditingMode ) {
+					this._showSourceEditing();
+					this._disableCommands();
+				} else {
+					this._hideSourceEditing();
+					this._enableCommands();
+				}
+			} );
+
+			this.on( 'change:isEnabled', ( evt, name, isEnabled ) => this._handleReadOnlyMode( !isEnabled ) );
+
+			this.listenTo( editor, 'change:isReadOnly', ( evt, name, isReadOnly ) => this._handleReadOnlyMode( isReadOnly ) );
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	afterInit() {
+		const editor = this.editor;
+
+		const pluginNamesToWarn = [
+			'RealTimeCollaborativeEditing',
+			'CommentsEditing',
+			'TrackChangesEditing',
+			'RevisionHistory'
+		];
+
+		// Currently, the basic integration with Collaboration Features is to display a warning in the console.
```

Comment:
I'd add a section about the plugin limitation in the documentation. As for now, the code could be translated into words and copied to the feature guide.

Reply:
Currently, the feature guide contains a warning, that this plugin will work out-of-the-box only for classic editor and it may not work with CF. IMHO this is sufficient for a feature guide. Integrator will be also notified in the browser's console in case his/her editor setup contains specific plugins from CF. I wouldn't add these CF plugin names to the feature guide as it would be too specific/detailed.